planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+)
+
+type outerJoinEliminator struct {
+	hasDistinct int
+	cols        [][]*expression.Column
+	schemas     []*expression.Schema
+}
+
+// tryToEliminateOuterJoin will eliminate outer join plan base on the following rules
+// 1. outer join elimination: For example left outer join, if the parent only use the
+//    columns from left table and the join key of right table(the inner table) is a unique
+//    key of the right table. the left outer join can be eliminated.
+// 2. outer join elimination with distinct: For example left outer join. If the parent
+//    only use the columns from left table with 'distinct' label. The left outer join can
+//    be eliminated.
+func (o *outerJoinEliminator) tryToEliminateOuterJoin(p *LogicalJoin) LogicalPlan {
+	if len(p.children) != 2 {
+		return nil
+	}
+	switch p.JoinType {
+	case LeftOuterJoin:
+		return o.doEliminate(p, 1)
+	case RightOuterJoin:
+		return o.doEliminate(p, 0)
+	default:
+		return nil
+	}
+}
+
+func (o *outerJoinEliminator) doEliminate(p *LogicalJoin, isLeft int) LogicalPlan {
+	// outer join elimination with distinct
+	if o.hasDistinct > 0 {
+		cols := o.cols[o.hasDistinct-1]
+		allColsInSchema := true
+		for _, col := range cols {
+			if !p.children[1^isLeft].Schema().Contains(col) {
+				allColsInSchema = false
+				break
+			}
+		}
+		if allColsInSchema == true {
+			return p.children[1^isLeft]
+		}
+	}
+
+	// outer join elimination without distinct
+	// first, check whether the parent's schema columns are all in left or right
+	if len(o.schemas) < 1 {
+		return nil
+	}
+	for _, col := range o.schemas[len(o.schemas)-1].Columns {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := p.children[1^isLeft].Schema().FindColumn(columnName); c == nil {
```

Comment:
we should check whether any column of the inner child is used by the parent of the join. Actually, the following outer join can also be eliminated:
```sql
create table t1(a bigint, b bigint, c bigint, unique key idx(a, b));
create table t2(a bigint, b bigint, c bigint);

select sum(t2.c) from t2 left join t1 on t2.a = t1.a and t2.b = t1.b;
```

Reply:
Pls consider the following case:
```
table t1:
+------+------+------+
| a    | b    | c    |
+------+------+------+
|    1 |    1 |    1 |
|    1 |    2 |    1 |
|    2 |    2 |    1 |
| NULL | NULL |    1 |
| NULL | NULL |    1 |
+------+------+------+
table t2:
+------+------+------+
| a    | b    | c    |
+------+------+------+
|    1 |    1 |    1 |
| NULL | NULL |    1 |
| NULL | NULL |    1 |
+------+------+------+
```
In this case, if we eliminate left outer join. The result is not true.